tests/callbacks/test_quantization.py

```diff
@@ -139,3 +154,91 @@ def test_quantization_triggers(tmpdir, trigger_fn: Union[None, int, Callable], e
     trainer.fit(qmodel, datamodule=dm)
 
     assert qcb._forward_calls == expected_count
+
+
+def _get_observer_enabled(fake_quant: FakeQuantizeBase):
+    # ``torch.quantization.FakeQuantize`` checks ``observer_enabled[0] == 1``.
+    return fake_quant.observer_enabled[0] == 1
+
+
+@pytest.mark.parametrize(
+    "observer_enabled_stages",
+    [("train", "validate", "test", "predict"), ("train",), ("validate",), ("test",), ("predict",), ()],
+)
+@RunIf(quantization=True)
+def test_quantization_disable_observers(tmpdir, observer_enabled_stages):
+    """Test disabling observers"""
+    qmodel = RegressionModel()
+    qcb = QuantizationAwareTraining(observer_enabled_stages=observer_enabled_stages)
+    trainer = Trainer(callbacks=[qcb], default_root_dir=tmpdir)
+
+    # Quantize qmodel.
+    qcb.on_fit_start(trainer, qmodel)
+    fake_quants = list(module for module in qmodel.modules() if isinstance(module, FakeQuantizeBase))
+    # Disable some of observers before fitting.
+    for fake_quant in fake_quants[::2]:
+        fake_quant.disable_observer()
+
+    for stage, on_stage_start, on_stage_end in [
+        ("train", qcb.on_train_start, qcb.on_train_end),
+        ("validate", qcb.on_validation_start, qcb.on_validation_end),
+        ("test", qcb.on_test_start, qcb.on_test_end),
+        ("predict", qcb.on_predict_start, qcb.on_predict_end),
+    ]:
+        before_stage_observer_enabled = torch.as_tensor(list(map(_get_observer_enabled, fake_quants)))
+
+        on_stage_start(trainer, qmodel)
+        expected_stage_observer_enabled = torch.as_tensor(
+            before_stage_observer_enabled if stage in observer_enabled_stages else [False] * len(fake_quants)
+        )
+        assert torch.equal(
+            torch.as_tensor(list(map(_get_observer_enabled, fake_quants))), expected_stage_observer_enabled
+        )
+
+        on_stage_end(trainer, qmodel)
+        assert torch.equal(
+            torch.as_tensor(list(map(_get_observer_enabled, fake_quants))), before_stage_observer_enabled
+        )
+
+
+@RunIf(quantization=True)
+def test_quantization_val_test_predict(tmpdir):
```

Comment:
what is the meaning of this test, as I see you are running twice almost the same as default is the only train...
rather make assert on how many times the observer is called, probably some mock. @awaelchli?

Reply:
For most of the users, maybe the default quantization aware training should not be affected by validation, testing and predicting. Users will get the same trained model no matter validation, testing and predicting are performed or not.

There are two models in the test:

* `val_test_predict_qmodel` will be used by training, validation, testing and predicting.
* `expected_qmodel` will be used by training only.

The parameters of two trained models should be the same.

Reply:
yes, but there is not much reason why they shall be different as you do the same training steps or do I miss something?

Reply:
The old `QuantizationAwareTraining` will fail the test as the observers are not disabled during validating, testing and predicting. The new `QuantizationAwareTraining` should disable/enable observers properly and restore the fake-quantization modules' states after the sanity check to pass the test.

Reply:
so lest for `expected_qmodel` set sanity batches 0 and empty validation dataloader

Reply:
Thanks for the clarification !